File: internal/cli/list.go
package cli

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/cortex-ai/cortex-ai/internal/cli/output"
	"github.com/cortex-ai/cortex-ai/internal/memory"
	"github.com/cortex-ai/cortex-ai/internal/tui"
	"github.com/spf13/cobra"
)

var listCmd = &cobra.Command{
	Use:   "list",
	Short: "List all memories",
	Long: `List all memories with optional filtering.

Examples:
  cortex list
  cortex list --level semantic
  cortex list --include-obsolete`,
	RunE: runList,
}

var (
	listLevel           string
	listLimit           int
	listIncludeObsolete bool
	listJSON            bool
	listReverse         bool
)

func init() {
	listCmd.Flags().StringVarP(&listLevel, "level", "l", "", "Filter by level(s): working,episodic,semantic")
	listCmd.Flags().IntVar(&listLimit, "limit", 0, "Limit number of results")
	listCmd.Flags().BoolVar(&listIncludeObsolete, "include-obsolete", false, "Include obsolete memories")
	listCmd.Flags().BoolVar(&listReverse, "reverse", false, "Reverse sort order")
	listCmd.Flags().BoolVar(&listJSON, "json", false, "Output as JSON")

	rootCmd.AddCommand(listCmd)
}

func runList(cmd *cobra.Command, args []string) error {
	ctx := context.Background()

	// Initialize storage from config
	storageBackend, err := initStorage()
	if err != nil {
		return fmt.Errorf("failed to initialize storage: %w", err)
	}
	defer func() { _ = storageBackend.Close() }()

	var filterLevels []memory.MemoryLevel
	if listLevel != "" {
		for _, l := range strings.Split(listLevel, ",") {
			l = strings.TrimSpace(l)
			if !memory.IsValidLevel(l) {
				return fmt.Errorf("invalid level: %s", l)
			}
			filterLevels = append(filterLevels, memory.MemoryLevel(l))
		}
	}

	// List
	opts := memory.ListOptions{
		FilterLevels:    filterLevels,
		IncludeObsolete: listIncludeObsolete,
		Limit:           listLimit,
		Reverse:         listReverse,
	}

	memories, err := storageBackend.List(ctx, opts)
	if err != nil {
		return fmt.Errorf("failed to list memories: %w", err)
	}

	if listJSON {
		items := make([]output.ListItem, len(memories))
		for i, m := range memories {
			items[i] = output.ListItem{
				ID:        m.ID,
				Title:     m.Title,
				Level:     string(m.Level),
				Tags:      m.Tags,
				CreatedAt: m.CreatedAt,
				Obsolete:  m.Obsolete,
			}
		}
		jsonBytes, _ := json.MarshalIndent(items, "", "  ")
		_, _ = fmt.Fprintln(cmd.OutOrStdout(), string(jsonBytes))
		return nil
	}

	if len(memories) == 0 {
		_, _ = fmt.Fprintln(cmd.OutOrStdout(), tui.Subtle.Render("No memories found."))
		return nil
	}

	rows := make([][]string, len(memories))
	for i, m := range memories {
		obsoleteFlag := ""
		if m.Obsolete {
			obsoleteFlag = tui.Warning.Render("obsolete")
		}
		rows[i] = []string{
			tui.FormatLevel(string(m.Level)),
			tui.ShortID(m.ID),
			m.Title,
			m.CreatedAt.Format("2006-01-02 15:04"),
			obsoleteFlag,
		}
	}

	_, _ = fmt.Fprintln(cmd.OutOrStdout(), tui.RenderTable([]string{"LEVEL", "ID", "TITLE", "CREATED", "FLAGS"}, rows))
	_, _ = fmt.Fprintln(cmd.OutOrStdout(), tui.Subtle.Render(fmt.Sprintf("%d memor%s", len(memories), func() string {
		if len(memories) == 1 {
			return "y"
		}
		return "ies"
	}())))
	return nil
}

File: src/cortex/cli/list_.py
"""cortex list — list memories with filtering."""

from __future__ import annotations

from typing import Annotated

import typer
from rich.console import Console
from rich.table import Table

from cortex.cli._common import error, get_storage, memory_to_dict, print_json
from cortex.models.memory import MemoryLevel
from cortex.storage.base import ListOptions

console = Console()


def list_memories(
    level: Annotated[
        str, typer.Option("--level", "-l", help="Filter by level: working, episodic, semantic")
    ] = "",
    session: Annotated[str, typer.Option("--session", help="Filter by session ID")] = "",
    include_obsolete: Annotated[
        bool, typer.Option("--include-obsolete", help="Include obsolete memories")
    ] = False,
    limit: Annotated[int, typer.Option("--limit", help="Maximum number of results (0 = all)")] = 0,
    offset: Annotated[int, typer.Option("--offset", help="Number of results to skip")] = 0,
    reverse: Annotated[
        bool, typer.Option("--reverse", help="Reverse sort order (oldest first)")
    ] = False,
    as_json: Annotated[bool, typer.Option("--json", help="Output as JSON")] = False,
) -> None:
    """List memories with optional filtering."""
    filter_level: MemoryLevel | None = None
    if level:
        try:
            filter_level = MemoryLevel(level)
        except ValueError:
            error(f"Invalid level {level!r}. Must be: working, episodic, semantic")

    storage = get_storage()

    opts = ListOptions(
        level=filter_level,
        session_id=session,
        include_obsolete=include_obsolete,
        limit=limit,
        offset=offset,
        reverse=reverse,
    )

    try:
        memories = storage.list(opts)
    except Exception as exc:
        error(str(exc))
    finally:
        storage.close()

    if as_json:
        print_json([memory_to_dict(m) for m in memories])
        return

    if not memories:
        typer.echo("No memories found.")
        return

    table = Table(title="Memories", show_lines=True)
    table.add_column("ID", style="dim", width=10)
    table.add_column("Level", width=8)
    table.add_column("Title", style="bold")
    table.add_column("Tags")
    table.add_column("Updated", width=12)

    for m in memories:
        level_val = m.level if isinstance(m.level, str) else m.level.value
        table.add_row(
            m.id[:8],
            level_val,
            m.title,
            ", ".join(m.tags) if m.tags else "",
            m.updated_at.strftime("%Y-%m-%d"),
        )

    console.print(table)
    typer.echo(f"\nTotal: {len(memories)}")
